pkg/scanner/validation_test.go

```diff
@@ -0,0 +1,543 @@
+				}
+			}
+		})
+	}
+}
+
+// Test expression validation with inputs
+func TestValidateCELExpressionWithInputs(t *testing.T) {
+	tests := []struct {
+		name         string
+		expression   string
+		declarations []*expr.Decl
+		wantValid    bool
+		wantError    ValidationErrorType
+	}{
+		{
+			name:       "valid with declared variable",
+			expression: "pods.items.size() > 0",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "valid with multiple variables",
+			expression: "namespaces.items.all(ns, pods.items.exists(pod, pod.metadata.namespace == ns.metadata.name))",
+			declarations: []*expr.Decl{
+				decls.NewVar("namespaces", decls.Dyn),
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "undeclared variable in expression",
+			expression: "deployments.items.all(d, d.spec.replicas > 1)",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: false,
+			wantError: ValidationErrorTypeUndeclaredReference,
+		},
+		{
+			name:       "valid CEL macros - all",
+			expression: "pods.items.all(pod, pod.spec.containers.all(c, c.image != ''))",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "valid CEL macros - exists",
+			expression: "pods.items.exists(pod, pod.metadata.labels.app == 'test')",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "valid CEL macros - filter",
+			expression: "pods.items.filter(pod, pod.status.phase == 'Running').size() > 0",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+	}
+
+	validator := NewRuleValidator(nil)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			issues := validator.ValidateCELExpression(tt.expression, tt.declarations)
+
+			if tt.wantValid {
+				if len(issues) > 0 {
+					t.Errorf("Expected valid expression, got issues: %v", issues)
+				}
+			} else {
+				if len(issues) == 0 {
+					t.Errorf("Expected validation errors, but got none")
+				} else if tt.wantError != "" && issues[0].Type != tt.wantError {
+					t.Errorf("Expected error type %s, got %s", tt.wantError, issues[0].Type)
+				}
+			}
+		})
+	}
+}
+
+// Test CompileCELExpression public function
+func TestCompileCELExpression(t *testing.T) {
+	tests := []struct {
+		name       string
+		expression string
+		inputs     []Input
+		wantError  bool
+		errorMsg   string
+	}{
+		{
+			name:       "valid expression with inputs",
+			expression: "deployments.items.all(d, d.spec.replicas >= 2)",
+			inputs: []Input{
+				&InputImpl{
+					InputName: "deployments",
+					InputType: InputTypeKubernetes,
+				},
+			},
+			wantError: false,
+		},
+		{
+			name:       "expression with undeclared variable",
```

Comment:
One thing I like to do with test names is to include the outcome as part of the name. Here I might use:

```go
name:       "expression with undeclared variable fails to compile",
```

Which reinforces the behavior we're looking for in the test name - without having to jump to line 223 to see we're looking for an error. I find it a subtle thing that makes it easier to understand what the tests should do when you're running them.